### ISSUE TITLE
nullable subjson

### ISSUE THREAD

**Issue by Kleinast** [coduo/php-matcher#133]
Hi,

Is there a way to match that json property can be an object matching my matcher AND can also be null: 

```
{
  "id": "@integer@",
  "title": "@string@",
  "image": {
    "title": "@string@",
    "alt": "@string@",
    "url": "@string@.isUrl()"
  },
  "content": "@string@",
  "description": "@string@",
  "cookTime": "@integer@",
  "preparationTime": "@integer@",
  "coverNumber": "@integer@",
  "ingredients": "@string@",
  "type": "recipe"
}
```

But my image object can be null like this:

```
{
  "id": "@integer@",
  "title": "@string@",
  "image": null,
  "content": "@string@",
  "description": "@string@",
  "cookTime": "@integer@",
  "preparationTime": "@integer@",
  "coverNumber": "@integer@",
  "ingredients": "@string@",
  "type": "recipe"
}
```

Thanks for your help

**Comment by norzechowicz** [comment]
Hey, sorry at this moment it's impossible. But I think it should be possible to create something like:

```
{
   "image":"@array@.oneOf(optional(), json(...))"
}
```

This would require to create `json` expander that could take any valid json pattern and compare it with value.

**Comment by norzechowicz** [comment]
Hey, I'm happy to let you know that after I merged #173 following code will work just like you expected:

```php
(new MatcherFactory())->match(
  '{"image":null}',
  '{"image":"@null@||@json@.match({\"url\":\"@string@.isUrl()\"})"}'
);

(new MatcherFactory())->match(
  '{"image":{"url":"http://image.com"}}',
  '{"image":"@null@||@json@.match({\"url\":\"@string@.isUrl()\"})"}'
);

(new MatcherFactory())->match(
  '{"image":{"url":"http://image.com"}}',
  '{"image":"@json@.oneOf(optional(), match({\"url\":\"@string@.isUrl()\"}) )"}'
);
```

Feel free to use `dev-master` version in order to test it or wait for 4.0 release (that is coming soon!)

**Comment by Kleinast** [comment]
Awsome ! thx @norzechowicz
